docs/replication.md

```diff
@@ -0,0 +1,145 @@
+- Sharing : The continuous act of syncing change to a subset of the cozy documents and files to and from another user's cozy.
+- Backup : The regular but not continuous act of syncing the cozy data to
+
+
+CouchDB replication is a well-understood, safe and stable algorithm ensuring replication between two couchdb databases.
+Rsync is a well-understood, safe and stable algorithm to synchronize files hierarchy between two hosts.
+
+**Moving** the cozy is similar to a one time **Backup**
+
+# Files synchronization
+
+Replication of too heavy database (with a lot of attachments) has been, in cozy experience, the source of some bugs. To avoid that (and some other issues), in cozy-stack, the attachments are stored outside of couchdb.
+
+This means we need a way to synchronize files in parallel to couchdb replication.
+
+This could be done with a rsync-like mechanism: Rsync essentially works by computing a signature of the files on target, sending this signature to source. On source, the file signature is used to compute a diff and then send only the changed chunks of the file to target. The chunks are stored on a tmp file on disk, so subsequent rsync does not need to resend the whole file.
+
+The difficulty will be in maintaining synchronization of the couchdb & vfs state while synchronizing between hosts with two different replication algorithm. The MD5 stored in couchdb should help to ensure correctness. The already-planed *update-couchdb-to-vfs-state* function too.
+
+We might need a zsync route too.
+
+Some useful packages:
+- Pure go rsync implementation : https://godoc.org/github.com/smtc/rsync
+- Go bindings for librsync : https://github.com/silvasur/golibrsync
+- A package to sync folder in FS or Swift: http://rclone.org/ (not rsync based, always whole file transfer)
+
+
+
+# Couchdb replication & limitation
+
+Couchdb 2 replication protocol is described [in details here ](http://docs.couchdb.org/en/2.0.0/replication/protocol.html)
+
+
+## Quick summary
+
+1. The replicator figures out what was the last sequence number *lastseqno* which was correctly synced from the source database, using a `_local/:replicationid` document.
+2. The replicator `GET :source/changes?since=lastseqno&limit=batchsize` and identify which docs have changed (and their *open_revs*).
+3. The replicator `POST :target/_revs_diff` to get a list of all revisions of the changed documents the target does not know.
+4. The replicator `GET :source/:docid?open_revs=['rev1', ...]` several times to retrieve the missing documents revisions.
+5. The replicator `POST :target/_bulk_docs` with these documents revisions.
+6. Store the new last sequence number
+
+Repeat 2-5 until there is no more changes.
+
+## Details
+
+- In step 5, the replicator can also attempt to `PUT :target/:docid` if doc are too heavy, but this should not happens in cozy-stack considering there wont be attachment in couchdb.
```

Comment:
On the things I'd do differently with emails if I were to start a new, is to consider the content as a text/plain or text/html attachment. It is really inefficient to pass the  whole content just to see the messages list.